### DISCUSSION TITLE
Suggestion: view diffs while writing commit message

### DISCUSSION THREAD

**Discussion by canton7** [jesseduffield/lazygit#4076] [Ideas]
For complex commits, it's useful to be able to browse the list of files and their diffs while writing the commit message, to remind myself what's been included in the commit.

For bonus points, it's useful to be able to stage files while writing the commit message: you can use it as a checklist to make sure that you haven't missed anything.

`git gui` handles this nicely: there's a "commit message" box at the bottom of the screen which you can dip in and out of while doing other work, then an option to commit with that message:

![image](https://github.com/user-attachments/assets/e1a15a6b-709d-4f21-9a45-ed4df4009464)

When you opt to amend the previous commit, the box pre-populates with the previous commit message, which gives a nice way to edit the previous commit's contents and message at the same time.

A similar flow in lazygit could have the "commit" option focus a pane in the lower right, rather than having a dialog box pop up, perhaps? This pane could be visible while it has content, and hidden otherwise. "esc" returns focus back to the file pane, and "enter" commits as normal.

I appreciate this is probably a fair chunk of work, and people seem happy with the commit dialog as it is. If there's general support, I'd be happy to have a stab at implementing this.

Thanks!

**Comment by mark2185** [comment]
Alternative solution: set the `commit.verbose` option in your `gitconfig` to `true` and then commit using an external editor by pressing `C` instead of `c`.

For me it brings up `vim` with what you're describing.

**Reply by canton7** [reply]
That's true. It doesn't like you easily flick between files though, in the same way that you can flick through the file browser pane in lazygit. And you're constantly jumping up and down the edit window, which is a bit annoying: you can't see the message and diffs at the same time.

**Reply by goyalyashpal** [reply]
> set the `commit.verbose` option in your `gitconfig` to `true` and then commit using an external editor by pressing `C` instead of `c`.

yeah, i was missing the `git commit -v`.  is it possible to trigger verbose editor ad-hoc when wanted, rather than changing at configuration level?

**Reply by canton7** [reply]
I guess if you normally commit using the built-in editor (`c`), then only use the external editor (`C`) when you want to commit verbose, that would work? Or set up a custom shortcut to run `git - c commit.verbose=true commit` etc?

**Comment by stefanhaller** [comment]
I was a long-time user of git gui in a previous life (and then of Fork later, which has a very similar UI), so I can relate to what you are suggesting. However, I do agree that it would be a lot of work, and I'm not totally sure it's worth the effort.

But then, lazygit isn't that far away already. It remembers an unfinished commit message, so what I often do is begin writing a message, hit Esc to close the editor to browse the diffs some more, then hit `c` again to continue where I left off. It doesn't remember the last writing position, but hitting tab takes me to the end of the message body, which is usually where I want to continue typing. To me, this workflow almost qualifies as "good enough", I'd say.

**Reply by canton7** [reply]
> It remembers an unfinished commit message

Ooh, now *that* I hadn't realised, thanks! You're right, that does cover most of this use-case. The only difference is being able to see the commit message while browsing files, but I'm not convinced that's a particularly compelling use-case either. 

I'm happy to close this off, I think.
